File: LeetCode/remote/binary_tree_level_order_traversal.cs
//  https://leetcode.com/problems/binary-tree-level-order-traversal/
//  https://leetcode.com/submissions/detail/59043589/
//
//  Submission Details
//  34 / 34 test cases passed.
//      Status: Accepted
//      Runtime: 648 ms
//          
//          Submitted: 0 minutes ago

public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        var result = new List<IList<int>>();
        if (root == null)
        {
            return result;
        }
    
        result.Add(new List<int> { root.val });
        var left = LevelOrder(root.left);
        var right = LevelOrder(root.right);
        for (var i = 0; i < Math.Max(left.Count, right.Count); i++)
        {
           var level = new List<int>();
           if (i < left.Count)
           {
               level.AddRange(left[i]);
           }
    
            if (i < right.Count)
            {
                level.AddRange(right[i]);
            }
            
            result.Add(level);
        }
    
        return result;        
    }

    //  https://leetcode.com/submissions/detail/65154004/
    //
    //  Submission Details
    //  34 / 34 test cases passed.
    //      Status: Accepted
    //      Runtime: 544 ms
    //          
    //          Submitted: 1 minute ago
    public IList<IList<int>> LevelOrder(TreeNode root) {
        if (root == null) {
            return new List<IList<int>>();
        }
        
        var left = LevelOrder(root.left);
        var right = LevelOrder(root.right);
        return new List<IList<int>> { new [] { root.val }.ToList() }
               .Concat(left.Zip(right, (x, y) => x.Concat(y).ToList()))
               .Concat(left.Skip(right.Count()))
               .Concat(right.Skip(left.Count()))
               .ToList<IList<int>>();
    }
}

File: LeetCode/remote/binary_tree_level_order_traversal.rb
#   https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
#   Given a binary tree, return the zigzag level order traversal of its nodes' values.
#   (ie, from left to right, then right to left for the next level and alternate between).
#   https://leetcode.com/submissions/detail/56864436/
#
#   Submission Details
#   33 / 33 test cases passed.
#       Status: Accepted
#       Runtime: 92 ms
#           
#           Submitted: 0 minutes ago
#   You are here!
#   Your runtime beats 0.00% of rubysubmissions.
def zigzag_level_order(root, hash = nil, level = 0)
    hash ||= []
    return hash if root.nil?
    hash[level] ||= []
    if level & 1 == 0
        hash[level].push(root.val)
    else
        hash[level].unshift(root.val)
    end
    
    zigzag_level_order(root.left, hash, level + 1) if root.left
    zigzag_level_order(root.right, hash, level + 1) if root.right
    hash
end
